### PR TITLE
Add extra_requires to reduce dependencies

### PR DESCRIPTION
**Pull request by RunningLeon**
## Motivation

Requested by https://github.com/InternLM/lmdeploy/issues/459#issuecomment-1733358928

## Modification

split dependencies in `requirements.txt` into multiple parts and support pip install with extras:
- `requirements/build.txt`: 
- `requirements/runtime.txt`, # including`gradio`
- `requirements/lite.txt`
- `requirements/serve.txt`

*Note `tritonclient[all]` is removed and users have to install it manually, which is mentioned in the document.*

Users can install lmdeploy with extra_requires, eg:  `pip install lmdeploy[all]`

- `all`: equals to `pip install -r requirements/build.txt -r requirements/runtime.txt -r requirements/lite.txt -r requirements/serve.txt`
- `lite`: equals to `pip install -r requirements/runtime.txt -r requirements/lite.txt`
- `serve`: equalts to `pip install -r requirements/runtime.txt -r requirements/serve.txt`

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: requirements/runtime.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+fire
+gradio
```
Comment (lvhan028) [inline requirements/runtime.txt:2]: I think it's better to put gradio into serve.txt, @AllentDan what's your opinion?

File: setup.py [hunk 2]
```diff
@@ -133,7 +133,14 @@ def gen_packages_items():
               'lmdeploy': lmdeploy_package_data,
           },
           include_package_data=True,
-          install_requires=parse_requirements('requirements.txt'),
+          setup_requires=parse_requirements('requirements/build.txt'),
+          tests_require=parse_requirements('requirements/test.txt'),
+          install_requires=parse_requirements('requirements/runtime.txt'),
+          extras_require={
+              'all': parse_requirements('requirements.txt'),
```
Comment (lvhan028) [inline setup.py:140]: So, it is one of the following, right?
- pip install lmdeploy[all]
- pip install lmdeploy[lite]
- pip install lmdeploy[serve]

Comment (lvhan028) [reply setup.py:140]: Can we use `pip install lmdeploy`?

Comment (RunningLeon) [reply setup.py:140]: yes. we can install like:
- `pip install lmdeploy` or `pip install -e .`, which runs `pip install -r requirements/runtime.txt`
- `pip install lmdeploy[all]` or `pip install -e .[all]`, which runs `pip install -r requirements/runtime.txt -r requirements/lite.txt -r requirements/serve.txt`
- `pip install lmdeploy[lite]` or `pip install -e .[lite]`, which runs  `pip install -r requirements/runtime.txt -r requirements/lite.txt`
- `pip install lmdeploy[serve]` or `pip install -e .[serve]`, which runs  `pip install -r requirements/runtime.txt -r requirements/serve.txt`

File: requirements/runtime.txt (hunk 1 above)
Comment (AllentDan) [reply requirements/runtime.txt:2]: Agree.

File: README.md [hunk 3]
```diff
@@ -100,7 +100,15 @@ And the request throughput of TurboMind is 30% higher than vLLM.
 Install lmdeploy with pip ( python 3.8+) or [from source](./docs/en/build.md)
 
 ```shell
-pip install lmdeploy
+pip install lmdeploy[all]
```
Comment (AllentDan) [inline README.md:103]: what will happen if there is no `[all]`, only `pip install lmdeploy`?

File: README.md [hunk 4]
```diff
@@ -100,9 +100,16 @@ And the request throughput of TurboMind is 30% higher than vLLM.
...
```
Comment (AllentDan) [inline README.md:103]: can we add a description of `pip install lmdeploy`?

**Review by AllentDan — APPROVED** [review]
LGTM

File: README.md (hunk 4 above)
Comment (RunningLeon) [reply README.md:103]: sure. updated

File: requirements/lite.txt [hunk 5]
```diff
@@ -0,0 +1,2 @@
+accelerate
```
Comment (AllentDan) [inline requirements/lite.txt:1]: Shall we add version control here? @lvhan028

Comment (lvhan028) [reply requirements/lite.txt:1]: Problem solved after I updated accelerate to the latest version

**Comment by lvhan028** [comment]
PR #605 brings in `flash-attn`

**Comment by RunningLeon** [comment]
> PR #605 brings in `flash-attn`

@lvhan028 Shall I add `flash-atten` to `lite.txt` in this PR? Do we have version restriction for this package?

**Comment by lvhan028** [comment]
Qwen brought too many dependencies. We probably need a speicial depdency file for qwen

File: README.md [hunk 6]
```diff
@@ -100,9 +100,17 @@ And the request throughput of TurboMind is 30% higher than vLLM.
...
```
Comment (lvhan028) [inline README.md:103]: 建议在这里使用 pip install lmdeploy 保持文档历史上的一致性。
在下文中，使用到serving的地方，可以单独再安装对应的依赖包

**Review by lvhan028 — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (RunningLeon) [reply README.md:103]: ok

**Comment by lvhan028** [comment]
> > PR #605 brings in `flash-attn`
> 
> @lvhan028 Shall I add `flash-atten` to `lite.txt` in this PR? Do we have version restriction for this package?

Yes, to `lite.txt`. No version restrictions

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**